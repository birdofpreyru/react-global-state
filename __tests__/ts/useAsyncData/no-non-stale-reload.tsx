/** @jest-environment jsdom */

// This test checks that when up-to-date async data are present in the state,
// mounting the useAsyncData() hook related to those data does not trigger
// unnecessary reload.

import pretty from 'pretty';

import {
  type DestroyableHtmlElement,
  act,
  mount,
  timer,
} from 'jest/utils';

import {
  type AsyncDataEnvelopeT,
  newAsyncDataEnvelope,
  withGlobalStateType,
} from 'src';

type StateT = {
  key: AsyncDataEnvelopeT<string>;
};

const { GlobalStateProvider, useAsyncData } = withGlobalStateType<StateT>();

const loader = jest.fn(() => 'bad');

function Component() {
  const e = useAsyncData('key', loader);
  return <div>{e.data}</div>;
}

function Scene() {
  return (
    <GlobalStateProvider
      initialState={{
        key: newAsyncDataEnvelope('ok', {
          timestamp: Date.now(),
        }),
      }}
    >
      <Component />
    </GlobalStateProvider>
  );
}

let scene: DestroyableHtmlElement | undefined;

beforeEach(async () => {
  loader.mockClear();
  scene = mount(<Scene />);
  await act(() => timer(10));
});

it('does not reload data', () => {
  expect(loader).not.toHaveBeenCalled();
  expect(pretty(scene!.innerHTML)).toMatchSnapshot();
});
