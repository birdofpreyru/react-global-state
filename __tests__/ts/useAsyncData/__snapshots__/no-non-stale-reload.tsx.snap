// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`does not reload data 1`] = `
<div>
  <div>
    ok
  </div>
</div>
`;
